### PR TITLE
Fix operators that should be using IS NOT DISTINCT FROM rather than EQUALS

### PR DESCRIPTION
**Pull request by erichwang**
We have a few operators that should be using IS NOT DISTINCT FROM rather than EQUALS to provide the proper grouping semantics. This results in queries incorrectly grouping values that are NaN, as well as structural types that contain null elements.

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/gen/TestJoinCompiler.java
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/gen/TestJoinCompiler.java:114]: add some `nulls` and `NaNs` to `channel`?

File: presto-main/src/test/java/io/prestosql/operator/TestStreamingAggregationOperator.java
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/operator/TestStreamingAggregationOperator.java:109]: missing nulls

File: presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java:735]: please pick up https://github.com/trinodb/trino/pull/6462 as a prefix commit

**Review by sopel39 — APPROVED** [review]
let's wait for others (@dain, @martint  ) approval too. Should `positionEqualsPosition` and other methods be removed as they have strange semantics (nulls are equal, but `NaN` are not)

**Comment by sopel39** [comment]
per standard:
```
 If RANK is specified, then the rank of row R is defined as 1 (one) plus the number of rows that precede R and are not peers of R.
NOTE 26 — This implies that if two or more rows are not distinct with respect to the window ordering, then there will be one or more gaps in the sequential rank numbering.
```

**Comment by findepi** [comment]
Can you add query-based tests verifying correctness end-to-end as well?

**Comment by erichwang** [comment]
> Should positionEqualsPosition and other methods be removed as they have strange semantics (nulls are equal, but NaN are not)

@sopel39, we still need some of these because the equality semantics defined here are needed for equijoins, which use SQL equals, and not the distinct from shape. The nulls themselves are irrelevant for most of those use cases because they are assumed to be filtered out ahead of time for joins.

File: presto-main/src/main/java/io/prestosql/operator/PagesHashStrategy.java
Comment (erichwang) [reply presto-main/src/main/java/io/prestosql/operator/PagesHashStrategy.java:69]: Sorry, could you clarify this comment on what you think should be moved? I put the NOT DISTINCT variants directly below each of the EQUALS variants.

File: presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java
Comment (erichwang) [reply presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java:735]: Sorry, this might be kind of n00b, but what does it mean to pick up a prefix commit?

**Comment by sopel39** [comment]
> @sopel39, we still need some of these because the equality semantics defined here are needed for equijoins

Join use `io.prestosql.operator.PagesHashStrategy#positionEqualsRowIgnoreNulls` (they ignore nulls). It seems that every use of `positionEqualsPosition` (and other corresponding methods) is suspicious

File: presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java
Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java:735]: I mean to cherry pick it before other commits

**Review by findepi — CHANGES_REQUESTED** [review]
( https://github.com/trinodb/trino/pull/6472#issuecomment-752628920 )

File: presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java
Comment (erichwang) [reply presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java:735]: You mean integrate into this PR chain right?

Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/operator/TestWindowOperator.java:735]: Yup. Or rebase after I merge that PR

**Comment by sopel39** [comment]
Mind automation. There is error prone check failure

**Comment by erichwang** [comment]
> Mind automation. There is error prone check failure

Yea saw that, but this check failure doesn't make sense to me. Where is the protobuf coming from? How "error prone" are our error prone checks? =)

> Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project presto-spi: Resolution of annotationProcessorPath dependencies failed: Unable to get dependency information for com.google.protobuf:protobuf-java:jar:3.4.0: Failed to retrieve POM for com.google.protobuf:protobuf-java:jar:3.4.0: Could not transfer artifact com.google.protobuf:protobuf-java:pom:3.4.0 from/to central (https://repo.maven.apache.org/maven2): /home/runner/.m2/repository/com/google/protobuf/protobuf-java/3.4.0/protobuf-java-3.4.0.pom.part (No such file or directory)

**Comment by sopel39** [comment]
> Failed to retrieve POM for com.google.protobuf:protobuf-java:jar:3.4.0: Could not transfer artifact com.google.protobuf:protobuf-java:pom:3.4.0 from/to central (https://repo.maven.apache.org/maven2): /home/runner/.m2/repository/com/google/protobuf/protobuf-java/3.4.0/protobuf-java-3.4.0.pom.part (No such file or directory)

That is intermittent

**Comment by findepi** [comment]
> Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project presto-spi: Resolution of annotationProcessorPath dependencies failed: Unable to get dependency information for com.google.protobuf:protobuf-java:jar:3.4.0: Failed to retrieve POM for com.google.protobuf:protobuf-java:jar:3.4.0: Could not transfer artifact com.google.protobuf:protobuf-java:pom:3.4.0 from/to central (https://repo.maven.apache.org/maven2): /home/runner/.m2/repository/com/google/protobuf/protobuf-java/3.4.0/protobuf-java-3.4.0.pom.part (No such file or directory)

This looks similar to, but not the same as  https://github.com/trinodb/trino/issues/6414

Generally, the maven build needs to be retried in general on CI, as communication with Maven Central seems quite common cause of flakes.
It's likely we miss some more retries there.

Would you mind filing an issue?

**Comment by erichwang** [comment]
> Can you add query-based tests verifying correctness end-to-end as well?

Already added.

@findepi, any other comments?

**Comment by findepi** [comment]
> > Can you add query-based tests verifying correctness end-to-end as well?
> 
> Already added.

@erichwang thanks for adding a test for window. Did you also intend to add one for Join (also being changed here)?

**Comment by erichwang** [comment]
@findepi, this actually doesn't affect SQL joins (even though the logic is stored in something called JoinCompiler). The compiler class is very inaccurately named =), and is used in a lot of places.

File: core/trino-main/src/main/java/io/trino/operator/GroupedTopNRankAccumulator.java
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/GroupedTopNRankAccumulator.java:577]: commit message is too long

**Review by sopel39 — APPROVED** [review]

**Pull request merged**
